### PR TITLE
Fix solve_univariate_inequality

### LINKED ISSUE #25697: can not reduce log inequalities
```
>>> reduce_inequalities([log(x,3)<=2], x)
...
File .../sympy/solvers/inequalities.py:555, in solve_univariate_inequality.<locals>.valid(x)
    553     return expr.func(v, 0)
    554 # not comparable or couldn't be evaluated
--> 555 raise NotImplementedError(
    556     'relationship did not evaluate: %s' % r)

NotImplementedError: relationship did not evaluate: -2 + log(9)/log(3) <= 0
```
sympy version: 1.12. 

It seems that when `v` is almost zero, say `0.e-110`, `v.is_comparable` is False: 

```python
                  v = v.n(2)
                  if v.is_comparable:
                      return expr.func(v, 0)
                  raise NotImplementedError(
                       'relationship did not evaluate: %s' % r)
```

### PR DESCRIPTION
**Pull request by vitrun**
Improved Handling of Expressions Evaluating to Zero.

#### References to other Issues or PRs

Fixes #25697 

#### Brief Description of the Fix
This fix addresses the scenario where the variable `v` can be simplified to zero, such as in the case of `-2 + log(9)/log(3)`. To handle this situation, the code now utilizes `v.n(2, chop=True)` to ensure that `v` is evaluated as zero. 

#### Additional Comments
While this fix may not be considered the optimal solution, it effectively resolves the issue.

#### Release Notes

* solvers
  * Improved Handling of Expressions Evaluating to Zero.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
> This fix addresses the scenario where the variable `v` can be simplified to zero, such as in the case of `-2 + log(9)/log(3)`. To handle this situation, the code now utilizes `v.n(2, chop=True)` to ensure that `v` is evaluated as zero.

We can't use approximate numerical evaluation to determine that an expression is equal to zero. Only exact symbolic manipulation could prove that a zero expression is exactly zero.

**Comment by smichr** [comment]
You can use if `v.is_comparable or v.equals(0)` instead. It is ok to use n(2), but you can't ignore what it tells you via a residual real or imaginary part with no precision.

File: sympy/solvers/inequalities.py [hunk 1]
```diff
@@ -549,7 +549,7 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
+                    v = v.n(2, chop=True)
                     if v.is_comparable:
```
Comment (smichr) [inline sympy/solvers/inequalities.py:553]:
```suggestion
                    if v.is_comparable or v.equals(0):
```

File: sympy/solvers/inequalities.py [hunk 2]
```diff
@@ -549,8 +549,7 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
-                    if v.is_comparable:
+                    if v.is_comparable or v.equals(0):
                         return expr.func(v, 0)
```
Comment (smichr) [inline sympy/solvers/inequalities.py:553]:
```suggestion
                    if v.is_comparable:
                        return expr.func(v, 0)
                     if v.equals(0):
                        return expr.func(0, 0) 
```

File: sympy/solvers/tests/test_inequalities.py [hunk 3]
```diff
@@ -446,6 +446,10 @@ def test_issue_10671_12466():
         Interval.Lopen(6, 7)
 
 
+def test_issue_25697():
+    assert reduce_inequalities([log(x, 3) <= 2], x) === True
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:450]:
```suggestion
    assert reduce_inequalities([log(x, 3) <= 2], x) === True
    assert reduce_inequalities([log(x, 3) < 2], x) === False
```

File: sympy/solvers/inequalities.py [hunk 4]
```diff
@@ -549,9 +549,10 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
                     if v.is_comparable:
                         return expr.func(v, 0)
+                     if v.equals(0):
+                        return expr.func(0, 0)
```
Comment (oscarbenjamin) [inline sympy/solvers/inequalities.py:555]: We should not use `.equals` here.

Comment (smichr) [reply sympy/solvers/inequalities.py:555]: This is a higher level routine and use of simplification is not prohibited when needed to get a solution. Why do you think it should not be used?

File: sympy/solvers/inequalities.py [hunk 5]
```diff
@@ -549,9 +549,10 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
                     if v.is_comparable:
                         return expr.func(v, 0)
+                     if v.equals(0):
```
Comment (smichr) [inline sympy/solvers/inequalities.py:554]:
```suggestion
                    if v.equals(0):
```

File: sympy/solvers/inequalities.py [hunk 6]
```diff
@@ -549,9 +549,10 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
                     if v.is_comparable:
                         return expr.func(v, 0)
+                    if v.equals(0):
+                        return expr.func(0, 0)
```
Comment (smichr) [inline sympy/solvers/inequalities.py:555]:
```suggestion
                        return expr.func(0, 0)
```

File: sympy/solvers/tests/test_inequalities.py [hunk 7]
```diff
@@ -446,6 +446,11 @@ def test_issue_10671_12466():
...
+    assert reduce_inequalities([log(x, 3) < 2], x) === False
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:451]:
```suggestion
    assert reduce_inequalities([log(x, 3) < 2], x) == False
```

File: sympy/solvers/tests/test_inequalities.py [hunk 8]
```diff
@@ -446,6 +446,11 @@ def test_issue_10671_12466():
         Interval.Lopen(6, 7)
 
 
+def test_issue_25697():
+    assert reduce_inequalities([log(x, 3) <= 2], x) === True
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:450]:
```suggestion
    assert reduce_inequalities([log(x, 3) <= 2], x) == True
```

File: sympy/solvers/tests/test_inequalities.py [hunk 9]
```diff
@@ -446,6 +446,11 @@ def test_issue_10671_12466():
         Interval.Lopen(6, 7)
 
 
+def test_issue_25697():
+    assert reduce_inequalities([log(x, 3) <= 2], x) == True
+    assert reduce_inequalities([log(x, 3) < 2], x) == False
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:451]:
```suggestion
    assert reduce_inequalities([log(x, 3) <= 2], x) == (x <= 9) & (0 < x)
    assert reduce_inequalities([log(x, 3) < 2], x) == (x < 9) & (0 < x)
```

File: sympy/solvers/inequalities.py (hunk 4 above)
Comment (oscarbenjamin) [reply sympy/solvers/inequalities.py]: There are bigger problems here like the fact that there is no definition of what cases solve_univariate_inequality should be expected to handle or what outputs it should return. Attempting to work around those with simplify will only make those worse. In fact attempting to improve anything about these inequalities solvers without first addressing the problem of defining what they are even supposed to do is a mistake.

Comment (smichr) [reply sympy/solvers/inequalities.py]: The name implies that it should solve the inequality for a symbol if possible. This is an easy equality to solve but proving that 9 is a solution requires showing that `log(9)/log(3) == 2`. That can't happen with out `expand_log`. Constraints of the `A=B`  problem being acknowledged, we cannot refrain from returning a solution for something that is easily shown to be a solution just because it might fail in general (and return None). Nor should every routine that finds itself in the `A - B = 1e-126` sort of dead-end internally have to re-code the various simplifications that might be needed to show that the `A - B` actually *is* 0. There is nothing that can be simplified about the original expression `log(x, 3) <= 2`. Autoexpansion to `log(x)/log(3) <= 2` makes the substitution of 9 for `x` fail evaluation because the context (that the base is 3) is now lost. And even if we solve this problem at a higher level, the un-ending possibilities remain and `equals` is the last defense against repeating the heuristics throughout the solve mechanics. 

`equals` is not the right thing (as you have instructed) for things like `is_positive` which cannot risk getting trapped in a slow evaluation. It is the right thing for a `solve` routine (and the docstring indicates that it should return solutions for nonlinear inequalities.

File: sympy/solvers/tests/test_inequalities.py [hunk 10]
```diff
@@ -446,6 +446,11 @@ def test_issue_10671_12466():
         Interval.Lopen(6, 7)
 
 
+def test_issue_25697():
+    assert reduce_inequalities([log(x, 3) <= 2], x) == (x <= 9) & (0 < x)
+    assert reduce_inequalities([log(x, 3) < 2], x) == (x < 9) & (0 < x)
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:451]:
```suggestion
    assert reduce_inequalities([log(x, 3) <= 2], x) == And(x <= 9) & (S(0) < x)
    assert reduce_inequalities([log(x, 3) < 2], x) == (x < 9) & (S(0) < x)
```

File: sympy/solvers/inequalities.py (hunk 4 above)
Comment (oscarbenjamin) [reply sympy/solvers/inequalities.py]: > The name implies that it should solve the inequality for a symbol if possible

What does "solve" mean and what does "if possible" mean. The expected behaviour of this function is entirely unclear to me.

> we cannot refrain from returning a solution for something that is easily shown to be a solution just because it might fail in general (and return None

Why not?

Why can't the inequality be returned so that the user can call simplify if they want?

The documented specification of this function gives no answer as to what should be expected. The fact that things might fail in general is important: it would be much better if this function had a defined restricted scope and could be expected to "solve" all problems within that scope.

Comment (smichr) [reply sympy/solvers/inequalities.py]: > What does "solve" mean and what does "if possible" mean. 

This is the same question that is raised with `simplify`. `fu` comes the closest to saying what that means because they offer a complex series of simplification schemes to select an answer. In `simplify` we have a series of steps (which might be modified) that can be used to simplify expressions and possibly show that an expression is equal to zero. So for `sympy`, "solve if possible" means, "with the heuristics we employ we find a solution and return the evaluated solution if it can be coaxed to evaluation with the simplification methods. If not, we will either raise an error or return an unevaluated solution."

> Why can't the inequality be returned so that the user can call simplify if they want?

I hadn't considered returning `log(9)/log(3) <= 2` as a solution since I am usually thinking about selecting a solution to an equality rather than an inequality. In this case, that might be reasonable. If there is more than one solution then an `Or` would be returned as the unevaluated solution. I am not sure the same could be done with simplify.

**Comment by smichr** [comment]
@vitun, if you want to try a different approach you can open a new pr. In a case like this instead of checking for `valid` to evaluate, a conditional expression could be added. This particular solution would get added as `ConditionSet(x, log(9)//log(3) <= 2, {9})`.

**Comment by oscarbenjamin** [comment]
> This particular solution would get added as `ConditionSet(x, log(9)//log(3) <= 2, {9})`.

The return here is a Boolean rather than a Set. Complicated routines like `simplify`, `solve` etc are really not needed here:
```python
In [18]: log(x, 3) <= 2
Out[18]: 
log(x)    
────── ≤ 2
log(3)    

In [19]: log(x, 3)*log(3) <= 2*log(3)
Out[19]: log(x) ≤ 2⋅log(3)

In [20]: exp(log(x, 3)*log(3)) <= exp(2*log(3))
Out[20]: x ≤ 9
```

**Comment by smichr** [comment]
The ConditionSet will be converted to a relationional at the end of the routine.

> Complicated routines like simplify, solve etc are really not needed here:

The docstring advertises solving non-linear routines and it is reasonable to expect that if it can be solved, it will. When we `intergrate` we don't hold back, we attempt all known strategies (or internals to that effect).

If the user wants to isolate a non-symbolic generator they are free to do so
```python
>>> solve_univariate_inequality(log(x)/log(3)<=2,log(x))
(-oo < log(x)) & (log(x) <= 2*log(3))
```
But if they request a solution for x and sympy can give it, they should be able to get it without doing the rewriting on their own.
```python
>>> solve_univariate_inequality(5*x/(x**2+1)>=2,x)
(1/2 <= x) & (x <= 2)
>>>
```
There are lower-level routines that are restricted in scope, isolating up to a generator as in 
```python
>>> from sympy.solvers.solvers import _invert
>>> _invert(2*x-3,x)
(3/2, x)
>>> _invert(2*x**2-3,x)
(3/2, x**2)
>>>
```
One might argue that the full machinery of `solve` can be requested and that that is the better place to make the request. But there has been discussion of removing solving of inequalities from `solve`. The logical workhorse is `solve_univariate_inequality` and we should not limit its ability to return a solution (though we might want to let it use helpers that are more limited in scope). But even if it uses helpers, solving this log equation will require the use of simplification to prove the validity of the solution.

**Comment by oscarbenjamin** [comment]
> > Complicated routines like simplify, solve etc are really not needed here:
> 
> The docstring advertises solving non-linear routines and it is reasonable to expect that if it can be solved, it will.

Here we are trying to "solve" the inequality `log(x, 3) < 2` with the intention to get something in the form `a < x < c`.
```python
In [1]: solve_univariate_inequality(log(x, 3) <= 2, x)
...
NotImplementedError: relationship did not evaluate: -2 + log(9)/log(3) <= 0
```
Somehow the solver decides that what it needs to do is determine whether `-2 + log(9)/log(3) <= 0` and it attempts to do so using numerical evaluation which fails in this case because numerical evaluation cannot be used to prove that a zero expression is zero. You are suggesting that `simplify` could be used in order to prove that the expression is equal to zero:
```python
In [2]: simplify(-2 + log(9)/log(3))
Out[2]: 0

In [3]: simplify(-2 + log(9)/log(3) <= 0)
Out[3]: True
```
What I am saying is that something is seriously wrong with the whole algorithm or approach that `solve_univariate_inequality` uses if it is necessary to use simplify here at all. Rather than trying to fix this by adding a call to simplify the general approach should be changed. In this case it is trivial to solve this inequality. First multiply both sides by `log(3)` which is a positive number:
```python
In [4]: lhs, rhs = log(x, 3), 2

In [5]: lhs <= rhs
Out[5]: 
log(x)    
────── ≤ 2
log(3)    

In [6]: lhs*log(3) <= rhs*log(3)
Out[6]: log(x) ≤ 2⋅log(3)
```
Now we need to invert the log using exp. The rhs is real but we need the lhs to be real as well. Requiring that `log(x)` be real means requiring that `x > 0`. With that assumption we can take `exp` of both sides since log and exp are inverses of one another over the reals and are both monotonically increasing functions:
```python
In [11]: exp(lhs*log(3)) <= exp(rhs*log(3))
Out[11]: x ≤ 9
```
This should be combined with the assumption `x > 0` which was required for the lhs to be real in the first place and that gives the expected final answer:
```python
In [12]: (0 < x) & (x <= 9)
Out[12]: x ≤ 9 ∧ x > 0
```

> But even if it uses helpers, solving this log equation will require the use of simplification to prove the validity of the solution.

Again it should not. The general approach should be chosen so that it is not necessary to prove the validity of the solution retrospectively: a solution should be constructed that is already known to be valid or is presented in such a way that its validity can be proven without the use of simplify. Using `simplify` does not work as a proof method because if `simplify` fails to simplify something then it proves nothing either way.

**Comment by smichr** [comment]
The method suggested for solving the inequality would essentially duplicate all the solve strategies so they can run simultaneously on both sides of the relational. SymPy already knows how to solve the equality. It also knows how to manipulate the expression so it is in a form that will evaluate favorably to show that the solution is valid. We should not be duplicating these capabilities to handle inequalities.

When a solution does not evaluate it is probably zero and the expression needs re-arrangement to show that. In this case, autoevaluation of `log(x,3)->log(x)/log(3)` destroys the context in which 9 could be recognized as a perfect power.

If you oppose the idea of simplification in general, do you oppose a series of targeted simplifications when the `a-b~=0` case is detected?

File: sympy/solvers/inequalities.py [hunk 11]
```diff
@@ -9,7 +9,7 @@
 from sympy.core.symbol import Symbol, Dummy
 from sympy.sets.sets import Interval, FiniteSet, Union, Intersection
 from sympy.core.singleton import S
-from sympy.core.function import expand_mul
+from sympy.core.function import expand_mul, expand_log
```
Comment (smichr) [inline sympy/solvers/inequalities.py:12]:
```suggestion
from sympy.core.function import expand_mul, expand_log
from sympy.functions.elementary.exponential import log
```

**Comment by oscarbenjamin** [comment]
> We should not be duplicating these capabilities to handle inequalities.

It only looks like duplication if you ignore the fact that every step needs to be handled differently in the case of inequalities rather than equations. Polynomials should be handled differently, functions like exp and sin should be handled differently. Even just multiplying the two sides of an inequality should be handled differently.

SymPy's inequality solvers are all broken and useless because of this misguided attempt to try to build an inequality solver on top of solve:
```python
In [1]: reduce_inequalities(sin(x) > 0, x)
Out[1]: 0 < x ∧ x < π
```
Here `solve` cannot represent the infinite solutions but `reduce_inequalities` can: it would be better if it simply returned the input unmodified.

Any code in SymPy that uses these solvers is broken and buggy because of basic problems like this:
```python
In [4]: x = symbols("x", real=True)

In [5]: integrate(abs(sin(x)), x)
Out[5]: 
⎧  cos(x)    for x ≤ 0
⎪                     
⎨2 - cos(x)  for x ≤ π
⎪                     
⎩cos(x) + 4  otherwise
```

**Comment by smichr** [comment]
I don't think any solver handles assumptions as they should (i.e. returning conditions on the solution when dividing by a symbolic expression that could be zero). That is the state of SymPy at present. 

> should be handled differently

While solving an inequality requires care during inversion, this care is taken, when solving as an equality, by substituting the solution into the original expression to see if it is valid or not, I am open to demonstration that this is a flawed approach (but do not accept the current issue of this PR as valid since I already argue that the "care" step happens with simplification instead of  as manipulation during the inversion process).

> solve cannot represent the infinite solutions

This is a known issue. Returning principle roots by other routines and (in this case) solutions valid within one period are current limitations. I made a suggestion of how that might be handled with a Subs-like mechanism on the mailing list. It is a known limitation. I would rather have a piece of the solution than no solution at all.

> basic problem like this

Could you please give me a hint of what I should see? Are you saying that the output is wrong because `integrate` relied on `solve`?

**Comment by oscarbenjamin** [comment]
> I don't think any solver handles assumptions as they should (i.e. returning conditions on the solution when dividing by a symbolic expression that could be zero)

Many functions do not return a result in a format that could represent this but `reduce_inequalities` does.

> While solving an inequality requires care during inversion, this care is taken, when solving as an equality, by substituting the solution into the original expression to see if it is valid or not

The whole approach of substituting back in and trying to check validity is a mistake. It is a mistake when solve does it (`check=True`) and it is also a mistake to do it here. Any approach that ends up with the code needing to decide whether or not some expression is zero should be avoided because deciding zero equivalence is a very hard problem and is in fact impossible in general. There are some situations where substituting in and evaluating numerically is reasonable but the goal must be to prove numerically that a known nonzero expression is positive or negative rather than trying to prove that a zero expression is zero.

> I would rather have a piece of the solution than no solution at all.

Maybe that reasoning makes sense for `solve` but it does not make sense for `reduce_inequalities` or `simplify` or `integrate` etc. If `solve` might return only a "piece" of the solution then it cannot be used in any situation where a known complete solution is needed. If `solve` does not return complete solution sets and also does not give any indication as to whether the solution set is complete then it cannot serve as a foundation for other functions without leading to incorrect results.

> Could you please give me a hint of what I should see? Are you saying that the output is wrong because `integrate` relied on `solve`?

The integration output is wrong because `_solve_inequality` is used. Also other examples are wrong because of the use of `reduce_inequalities` or other functions from `solvers/inequalities`:
```python
In [19]: x = symbols("x", real=True)

In [20]: p = abs(sin(x)).rewrite(Piecewise)

In [21]: p
Out[21]: 
⎧sin(x)   for sin(x) ≥ 0
⎨                       
⎩-sin(x)    otherwise   

In [22]: simplify(p) # incorrect
Out[22]: 
⎧sin(x)   for x ≥ 0 ∧ x ≤ π
⎨                          
⎩-sin(x)      otherwise
```
Looking at it the problem is not using `solve` but just using code that has the same limitations. It actually uses `solvify` but it sets the domain to `[0, 2*pi)` before even calling `solvify` to reduce the output to what `solve` would have returned anyway:
https://github.com/sympy/sympy/blob/6e347f05210519d5eef3f8b894ccf5c4d7029226/sympy/solvers/inequalities.py#L507-L519

**Comment by smichr** [comment]
> What I am saying is that something is seriously wrong with the whole algorithm or approach

> If solve does not return complete solution sets and also does not give any indication as to whether the solution set is complete then it cannot serve as a foundation for other functions without leading to incorrect results.

I agree that a better approach is needed.
We know that routines expecting a full solution should not use `solve`.
We know that the use of `simplify` on the faulty approach will solve the issue of this PR (and others not yet identified).
We cannot make something more broken (by using `chop` to prove/hope something is zero).
We can make a neutral change by doing one of two things:
  1. simplify the input equation -- and we already use expansion so there is precendent
  2. use simplification of `v` as I have proposed

That other routines use `solve_univariate_inequality` via `reduce_inequality` is a separate issue. (The docstring of `solve_univariate_inequality` already advertises that "the solution returned for trigonometric inequalities are restricted in its periodic interval.")

Could you not oppose solution 1 or 2?

**Comment by oscarbenjamin** [comment]
I would prefer never to use `simplify` in library code because its behaviour is too ill-defined.

In this case the right way to fix it is that we need an inequality solver that can invert simple inequalities but that needs to be implemented as an inequality solver rather than being built on top of an equation solver. Inverting something like `log(x) < 3 -> 0 < x < exp(3)` needs to be something that is explicitly implemented in the context of inequalities. The solver needs to apply the knowledge that we work with real sets and needs to know which functions can be considered invertible in an inequality sense over any given interval.

More generally I think that a lot of the problems like this in SymPy stem from using `solve` rather than something like Mathematica's `reduce` function:
https://reference.wolfram.com/language/ref/Reduce.html

The limitations of `solve` are that its return format must always be a finite list of explicit expressions. There are good reasons for sometimes wanting a finite list of explicit expressions and so in that situation it is good to have a function that can do that regardless of the necessary limitations that it must have.

In other situations though we need to know that we have not logically changed a statement or excluded certain cases etc. A function like `reduce` can be permitted to leave its input unmodified in cases where attempting to give explicit expressions is impossible or undesirable. The basic constraint for `reduce` is that the output should be logically equivalent to the input so it should always be safe to use.

The `reduce_inequalities` function should satisfy that same basic constraint: the output must be logically equivalent to the input. That means that it must never use any other function that cannot preserve that property. On the other hand a function like `solve_univariate_inequalities` *can* be built on top of something like `reduce`: first use `reduce` to simplify the inputs and then attempt to transform that into something explicit. In Mathematica you can think of these as being sort of equivalent when solving equations:
```mathematica
sol1 = Solve[eq, x]
sol2 = ToRules[Reduce[expr, x]]
```
Here `Reduce` does all the "solving" and then `ToRules` will turn that into substitution rules if possible.

SymPy's solveset is sort of like an attempt to make `reduce` but the requirement to use sets rather than booleans hobbles it because then it cannot properly express conditions about symbols that don't involve the variable of interest e.g.:
```python
>>> reduce(Eq(a*x, b), x)
                  ⎛    b        ⎞
(a = 0 ∧ b = 0) ∨ ⎜x = ─ ∧ a ≠ 0⎟
                  ⎝    a        ⎠
```
The degenerate case here (a=0 and b=0) cannot really be expressed as a set of values for `x` and neither can the condition `a != 0` for the generic case:
```python
In [5]: solveset(Eq(a*x, b), x)
Out[5]: 
⎧b⎫
⎨─⎬
⎩a⎭
```
There are sort of ways to express these things as sets but really it would be much better just to use booleans.

I think that if we had something like `reduce` then there would be no situations where it would seem reasonable to use `solve` in the internals of anything in SymPy. The `solve` function is nice for users but is not suitable as a foundation for anything else.

**Comment by smichr** [comment]
> I would prefer never to use simplify in library code because its behaviour is too ill-defined.

The solution I propose (via choice 2) is to use targeted simplification as needed to fill in the gap until `reduce` is written.

**Comment by oscarbenjamin** [comment]
> > I would prefer never to use simplify in library code because its behaviour is too ill-defined.
> 
> The solution I propose (via choice 2) is to use targeted simplification as needed to fill in the gap until `reduce` is written.

We don't need to write `reduce` to solve this case. What is needed is to focus on the right way to write the simple code that solves this problem. Adding in things like `simplify` to work around basic algorithmic failures makes it harder to come up with a proper solution in future because nothing else can reproduce the behaviour of `simplify`. Once we start depending on things like `simplify` the scope of each function becomes completely undefined.

We are talking about solving a trivial inequality here so I will repeat this once again:
```python
In [18]: log(x, 3) <= 2
Out[18]: 
log(x)    
────── ≤ 2
log(3)    

In [19]: log(x, 3)*log(3) <= 2*log(3)
Out[19]: log(x) ≤ 2⋅log(3)

In [20]: exp(log(x, 3)*log(3)) <= exp(2*log(3))
Out[20]: x ≤ 9
```
We should not even contemplate the use of `simplify` here: something else should be changed.

**Comment by smichr** [comment]
> a trivial inequality here

You stance is forcing `reduce` (or similar) to be written because the next failing expression might be `log(x + 1, 3) <= 2` or `log(x, 3) + 1 <= 2` which will require subtracting 1 at the right time; or `log(x**2, 3) <= 2` which will now require the `sqrt`. We already know how to do this for an equation. All we need is to help the verification of the solution with `expand_log` at the end (not `simplify`).

All the steps needed to solve the equation properly can either be applied at the beginning or the end, and doing so at either time makes doing so at the other unnecessary. Think of all the methods needed to invert `log(f(x), 3) <= 2`. Despite the many paths needed to invert until you have an expression that isolates `x`, if you solve `log(f(x), 3) - 2` they all will require verification of `log(9)/log(3) <= 2` which can be trivially shown to be true (and valid) with `expand_log`.

(But I am a little stumped by how solutions for `log(x**2 + 1, 3) <= 2` and `log(x + 1, 3) <= 2` can be found but not `log(x, 3) <= 2`.)

> We should not even contemplate the use of simplify here

We already agree on this point. I am suggesting targeted, well-defined simplification steps only. In this case, `expand_log`. 

I also agree that there is a better way but understand that there is no better plan in the works and would like to see an improvement in the current state in that it allows a valid solution to be returned.

**Comment by oscarbenjamin** [comment]
> You stance is forcing `reduce` (or similar) to be written because the next failing expression might be `log(x + 1, 3) <= 2` or `log(x, 3) + 1 <= 2` which will require subtracting 1 at the right time; or `log(x**2, 3) <= 2` which will now require the `sqrt`

Yes, these things should be written for the `reduce_inequalities` function. This is how it should handle simple inequalities that can be inverted one step at a time. 

Actually for `sqrt` I would rather say that it would need a way to handle polynomials. Those are a special well-defined case in which computing the roots and verifying signs can work.

> All we need is to help the verification of the solution with `expand_log` at the end (not `simplify`).

The same point applies here: `expand_log` might handle this one particular case but there will be others that it does not. Before long someone will replace the `expand_log` with `simplify` because it seems more powerful. Then forever the function will have no defined scope and will work heuristically in some cases and fail randomly for others.

An approach that leads to post-verification by attempting to simplify arbitrary expressions to zero will always necessarily fail in some cases. Using `simplify` to do this means that the scope of what can and cannot be solved will never be well defined.

It is not worth trying to patch up special cases when the general approach is clearly wrong. It is clear how to write the code that would solve this case in an unambiguously good way so the code to do that should be written instead.

The suggestion here is to add a shortcut as if that actually helps things but in the long term it harms things. The first problem is that none of these functions has a defined or documented scope: what is the set of problems that it should be expected to be able to solve?

When the scope is undefined that makes it very difficult for anyone to improve the code in future. SymPy has many solver functions that are in this situation where it is hard to improve them in a major way because it is completely unclear how to reproduce much of their existing behaviour after any modification of the code.

**Comment by oscarbenjamin** [comment]
@smichr if you want to fix this one issue using `expand_log` then I don't object to merging this.

I want to make clear though that this is further extending the wrong approach for implementing these solvers. It is really not worth trying to patch up individual cases when the code should be completely changed. In particular `reduce_inequalities` should be changed into something that can preserve exact logical equivalence between inputs and outputs because there is no reason in an API sense why it can't do that and the code that uses it already presumes that it does preserve logical equivalence (e.g. by using it in simplify). Making `reduce_inequalities` preserve logical equivalence though means that it should not be built on top of any of the existing solvers because none of them can provide that sort of guarantee.

File: sympy/solvers/solvers.py [hunk 12]
```diff
@@ -3144,7 +3144,7 @@ def _invert(eq, *symbols, **kwargs):
 
     >>> invert(sqrt(x + y) - 2)
     (4, x + y)
-    >>> invert(sqrt(x + y) - 2)
+    >>> invert(sqrt(x + y) + 2)
     (4, x + y)
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py:3148]: Why is this changed?

Comment (smichr) [reply sympy/solvers/solvers.py:3148]: it was a typo; the line above is identical

**Comment by smichr** [comment]
@oscarbenjamin , I really appreciate the time you have taken to hash this through. I understand your position and appreciate you being willing to entertain this one correction. I know that this will encourage "the next patch", perhaps, and end up looking like "simplify" at the end, but only heuristically patching up known trouble expressions.

As I was looking over it one more time before reading your final comments, I found the issue that is causing the validation step to fail and will push a new PR with that fix.

If anyone wants to make changes to solvers in the future, this could be a good place to send them for some commentary.
